### ISSUE TITLE
Convert Layout Icons from PNG to SVG

### ISSUE THREAD

**Issue by stpaultim** [backdrop/backdrop-issues#6487]
## Description of the need

It has been suggested that we convert layout icons from png to svg, now that we have svg icon library in core.
https://github.com/backdrop/backdrop-issues/issues/6480#issuecomment-2081480473

## Proposed solution

Convert png icons to svg. 

## Additional information

![image](https://github.com/backdrop/backdrop-issues/assets/3144571/118e1e67-0656-4fe4-a90c-82560c54634f)

**Comment by indigoxela** [comment]
There _might_ be a bit of a misunderstanding here.

We can easily just use svg instead of png for layout icons (they'd look less blurry and the regions less shifted).

However, as these are included as `img` tags, there's no way to recolor them. And changing the whole render process to display them either as bg-image as a mask, or to display inline svg (as markup) might be a breaking change for contrib. As that might require a change in the API.

**Comment by indigoxela** [comment]
Although I'm quite sure, this doesn't meet the intended feature, I provided a [PR](https://github.com/backdrop/backdrop/pull/4734) with SVG instead of PNG.

They look a bit cleaner, even when zooming in far with the browser. That's about it. :wink:

**Comment by indigoxela** [comment]
@stpaultim it was exactly that comment by @laryn that gave me the impression, that there must be a misunderstanding somewhere, in how svg as image (in img tag) works. :wink: 

Neither is it possible to recolor those by CSS, nor will building custom svg icons per flexi-layout be "trivial". The challenges of such an on-the-fly icon for custom layouts are quite the same, I guess. Be it SVG or raster image (via gd).

**Comment by klonos** [comment]
> Neither is it possible to recolor those by CSS, nor will building custom svg icons per flexi-layout be "trivial".

@indigoxela I have created SVG icons for the core layouts based on lines + `currentColor`, which allow them to behave the same as the Phosphor icons we use in the admin bar (inherit whatever color has been added via CSS for the text). I've also adjusted them to indicate the main region of each layout with a bit darker shade of the color (using a class within the SVG + CSS opacity):
<img width="747" alt="image" src="https://github.com/backdrop/backdrop-issues/assets/2423362/0d70531c-c582-4c4e-b604-1dd0033126c8">

Please let me know if you'd like a PR against yours, or if you prefer that I created a separate PR as an alternative.

As for automatically-generated icons for flexible layouts, there's already #4104, for which I've created a PR in the past based on native PHP functions like `imagecreate()`/`imagecolorallocate()`/`imagefilledrectangle()`/`imagepng()` (it was really close - just needed one specific thing figured out last I recall).

**Comment by klonos** [comment]
...the svg icon code looks like this:
```svg
<svg class="boxton" viewBox="0 0 100 150" xmlns="http://www.w3.org/2000/svg">
  <line class="header" x1="0" y1="7" x2="100" y2="7" />
  <line class="header" x1="0" y1="27" x2="100" y2="27" />
  <line class="main" x1="0" y1="75" x2="100" y2="75" />
  <line class="footer" x1="0" y1="123" x2="100" y2="123" />
  <line class="footer" x1="0" y1="143" x2="100" y2="143" />
  Sorry, your browser does not support inline SVG.
</svg>
```

**Comment by indigoxela** [comment]
@klonos I don't think we can change the core layout icons without considering flexible layouts in the same step, though.

Should I close my PR? It was more like a POC, anyway.

**Comment by klonos** [comment]
> I don't think we can change the core layout icons without considering flexible layouts in the same step, though.

I think we should. This change here will be very small (just replace the .png files with .svg ones + some minimal CSS changes), whereas handling automated generation of .svg thumbnails each time a flexible template is saved needs more code/login into it.

> And there are also some contrib layouts.

Yes. And some custom layouts as well I am assuming. These can continue to use whatever they want; .png/.jpeg/.gif or switch to .svg. Core should not be responsible for these - same as we are not responsible for the thumbnails shown for contrib themes in the project browser. We should be able to handle it though.

> Should I close my PR? It was more like a POC, anyway.

Up to you really. But it does sound like you were experimenting with the idea, so I'm going to unassign the issue from you and assign it to myself, as I would like to get this done. I hope you don't mind.

**Comment by klonos** [comment]
...perhaps as part of this issue/PR here we can create a .svg file to replace the current .png we are using for flexible templates, but yeah, I don't want to include the automatic generation here.

**Comment by indigoxela** [comment]
@klonos moved my PR out of the way. :wink:

**Comment by klonos** [comment]
I've hit a couple of snags:
- #6576
- #6575

I'll keep working on a PR for this, but have added `@todo` items for the things that need to be addressed properly.

**Comment by klonos** [comment]
Still working on it @izmeez 🔨 🔧  ...see how in my previous comment I said:
> I've hit a couple of snags:

**Comment by klonos** [comment]
I've also added the `documentation` label to this issue, because I would also like us to outline some basic steps to help people create such icons for their layout templates if they want. It's kinda simple and straight-forward if you start from an existing core-provided icon, and edit it in a text editor (some basic math and playing around might be required).

**Comment by klonos** [comment]
PR almost ready. ...as an added bonus, layout templates can now add extensionless icon names from the core-provided Phosphor set, and then these will be used. For example, changing the `taylor-flipped.info` file from this:
`preview = taylor-flipped.svg`
...to this:
`preview = lego`

Has the following effect:
![image](https://github.com/backdrop/backdrop-issues/assets/2423362/256d753c-78af-4b4c-bd87-8d4e88fb541e)

All flexible layouts are currently using this new "trick" to set their icon to `pencil-ruler`. I randomly chose that icon for now, but no need to waste time bikeshedding about it, since that will be short-lived (planning to work on #4104 next).

**Comment by klonos** [comment]
Here's the PR: https://github.com/backdrop/backdrop/pull/4769

**Comment by klonos** [comment]
...one set of tests is failing for php5 and php7, but I need to take a break. Sorry.

**Comment by klonos** [comment]
OK, texts fixed (another snag, related to our SVG support implementation and php8 specifically - I'll open a new issue for that), however I just noticed that in the PR sandbox the icons are not being generated/loaded 😞 ...my local dev is based on DDEV running php 8.3.6, whereas the sandbox is running 8.2.11 - both are on Apache/2.4.57 (Debian).

More digging required I guess. That's sad, because this is looking kinda awesome on my local.

Still setting this to `needs code review` and `needs testing` for people that can pull the PR and test things locally.

**Comment by jenlampton** [comment]
First @klonos - I love this change! Thank you for opening this :) 

The PR looks great, but I have concerns about styling the Rolph layout to indicate a `<main>` region when it doesn't have one. I think that's a bit deceiving. If the layout doesn't have one, the image should reflect that so there are no surprises down the road :)

**Comment by klonos** [comment]
@jenlampton yes, I agree. I'll work to address the feedback from the dev meting just then, including that problem with Rolph. ...what you need to check out though is this: 6461

**Comment by jenlampton** [comment]
@klonos oh, I thought you were indicating a `<main>` region, but you were instead indicating a default region -- that makes more sense actually. And since that's required every layout should have one :)

**Comment by klonos** [comment]
This is currently blocked on the following issues:
- #6577
- #6576
- #6575

I'm trying to unblock all these for the 1.28.1 release if possible.

**Comment by klonos** [comment]
One blocker down - two more to go 🙂 ...I've updated the PR to revert the changes that have now been merged into core as part of #6577.

**Comment by klonos** [comment]
...another blocker gone - only one left 🙂 ...updated the PR again, to remove the changes that are now in core with #6576, and rebased it as well while at it.
